### DISCUSSION TITLE
How to collect test coverage including untested files?

### DISCUSSION THREAD

**Discussion by DanilAgafonov** [nodejs/help#4564] [Q&A]
Is it possible to include untested files in coverage report?

Reason: I want to know how much code I have uncovered in my project overall. I want to see files for which I haven't created any tests and never imported them anywhere.

Example:
Let's say I have `src/index.js` file. And I run `node --test --experimental-test-coverage --test-coverage-include="src/**/*"` in the root of my project. My expectation: report should include `index.js` showing that zero lines covered with tests.

**Comment by nullswan** [comment]
Also interested ! I think this is related to https://github.com/nodejs/help/issues/4487

**Comment by MoLow** [comment]
Coverage is collected using v8's [takeCoverage](https://nodejs.org/api/v8.html#v8takecoverage) method, which means it is only aware of code that has actually run.
this might be a feature that is possible to implement but it is impossible with the current implementation

**Comment by hernandemonteiro** [comment]


**Reply by DanilAgafonov** [reply]
Thanks for a detailed reply.

I've actually been considering doing something like `import './src/index.js';` in test setup file. But I don't think this is option I want to use, because if I do that, the top-level code in these imported modules will be executed and marked as "test covered" (which I don't want). I want these never-imported modules be marked as untested/uncovered (I mean it should report that zero lines was executed during test in these files).

**Reply by rwalle** [reply]
Is this AI generated answer? It is almost completely irrelevant.

> Use `--test-coverage-include` to include all files (you are already doing this):

If they are already doing this, why do you even bring this up???

> If the files are still not included, consider importing them in a "setup" file to ensure they are loaded into the test environment.

How do you include these files? You need to come up with a list with precisely the files you want, which is difficult or at least very cumbersome to maintain. You need some sort of tool to automatically generate the list and make sure it is always up-to-date, possibly as part of a build step. And even so there are likely issues that could arise with this approach.

Mentioning nyc/istanbul and jest when people are clearly talking about the built-in test runner is also ridiculous.

This answer makes zero sense.

**Comment by lillybhagya** [comment]
**Install these packages**
_"jest": "^29.7.0",
"jest-environment-jsdom": "^29.5.0",
"babel-jest": "^29.7.0",_

The following configurations will work. Create a __tests__ folder at the root of your project, Ensure that all test files follow the *.test.js or *.spec.js naming convention.

**jest.config.js**
```
module.exports = {
  testEnvironment: 'jsdom',
  moduleFileExtensions: ['js', 'jsx', 'json'],
  moduleNameMapper: {
    '^@/(.*)$': '<rootDir>/src/$1',
    '\\.(css|less|scss)$': '<rootDir>/styleMock.js',
  },
  testMatch: ['<rootDir>/__tests__/**/*.(test|spec).{js,jsx}'],
  transformIgnorePatterns: ['<rootDir>/node_modules/'],
  collectCoverageFrom: [
    'src/**/*.{js,jsx}',
      '!src/assets/**',
      '!src/config/**',
      '!src/css/**',
      '!src/fonts/**',
  ],
  coverageReporters: ["text-summary"],
};
```

**Package.json**
_"test": "jest --config=jest.config.js --coverage"_

**Reply by avivkeller** [reply]
Just a reminder that this discussion is about the Node.js native test runner, not Jest.

**Comment by Errr0rr404** [comment]
Node’s built-in test runner coverage only instruments files that actually load during the run. Files that are never `import`/`require`’d usually stay missing unless you force them into the graph.

Options:

1. **Explicitly import every source file in a setup/bootstrap test** (simple but blunt):
   ```js
   // test/load-all.js
   import fs from "node:fs";
   import path from "node:path";
   function walk(dir) {
     for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
       const p = path.join(dir, e.name);
       if (e.isDirectory()) walk(p);
       else if (p.endsWith(".js")) import(pathToFileURL(p));
     }
   }
   ```
2. **Use c8 / istanbul / nyc** with `--all` / `all: true`, which can include files matching include globs even if unloaded:
   ```bash
   npx c8 --all --include "src/**/*.js" node --test
   ```
3. Keep `--test-coverage-include="src/**/*"` but remember it only filters what Node reports among *loaded* files — it does not by itself invent coverage for never-imported modules.

If you want “project-wide % including dead files”, `c8 --all` (or equivalent) is currently the least surprising approach.

**Comment by hanityx** [comment]
Node 26.7.0 added [`--test-coverage-include-all`](https://nodejs.org/api/cli.html#--test-coverage-include-all), which covers this case now.
 
  ```sh
  node --test --experimental-test-coverage \
    --test-coverage-include="src/**/*" \
    --test-coverage-include-all
  ```
  
It includes matching source files that were never loaded during the test run, 
so there's no need to import them from a setup file just to get them into the coverage report.

This was added after the earlier replies here, and it's still marked experimental.
